examples/nn.dx

```diff
@@ -0,0 +1,227 @@
+' # Neural Networks
+
+include "plot.dx"
+
+' ## NN Prelude
+
+
+def relu (input : Float) : Float =
+  select (input > 0.0) input 0.0
+
+instance [Add a, Add b] Add (a & b)
+  add = \(a, b) (c, d). ( (a + c), (b + d))
+  sub = \(a, b) (c, d). ( (a - c), (b - d))
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward:(param -> inp -> out) &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
+
+' ## Layers 
+
+' Dense layer
+
+def Dense (a:Type) (b:Type) : Type =
+  Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float))
+
+def dense (a:Type) (b:Type) : Dense a b =
+  {
+   forward = (\(weight, bias) x .
+               for j:b. (bias.j + sum for i :a. weight.i.j * x.i)),
+   init = arb
+  }
+
+' CNN layer
+
+def CNN (h:Int) (w:Int) (inc:Type) (outc:Type) (kw:Int) (kh:Int) : Type =
+  (Layer (inc=>Fin h=>Fin w=>Float)
+         (outc=>Fin h=>Fin w=>Float)
+         ((outc=>inc=>Fin kh=>Fin kw=>Float) &
+          (outc=>Float)))
+
+def conv2d (x:inc=>(Fin h)=>(Fin w)=>Float)
+           (kernel:outc=>inc=>(Fin kh)=>(Fin kw)=>Float) :
+     outc=>(Fin h)=>(Fin w)=>Float =
+     for o. for i.for j.
+         (i', j') = (ordinal i, ordinal j)
+         case (i' + kh) < h && (j' + kw) < w of
```

Comment:
cross-ref #333

Reply:
I find this pattern to be pretty common. I often want index arithmetic. My haskell brain says to add functions for `Maybe` index operators and then push around Maybe's with a bind. 

However I have no idea if that fits with the dex efficiency model. Conv is the first time I've hit dex efficiency issues. I might take a detour to try to learn more about how this gets turned into Cuda code.

Reply:
Yeah, wrapping arithmetic in helper functions is what I would recommend. See e.g. the `+|` functions defined in a bunch of examples for clamped-offsetting of indices.

Oh, are you using the CUDA backend? Dex doesn't use the GPU by default, because the codegen is not entirely fleshed out. You could try it by appending `--backend llvm-cuda`, but it looks like it regressed a little due to recent LLVM changes and fails to load the CUDA program onto the GPU...